### DISCUSSION TITLE
[Android] Render on a SurfaceTexture

### DISCUSSION THREAD

**Discussion by RockVampire** [google/filament#4295] [Q&A]
I'm trying to use filament on a SurfaceTexture on Android.
I write a method "attachTo" like the other "attachTo"  in the UIHelper 
`    public void attachTo(@NonNull SurfaceTexture surfaceTexture , int width, int height , SurfaceTexture.OnFrameAvailableListener listener) {
        if (attach(surfaceTexture)) {
            int format = isOpaque() ? PixelFormat.OPAQUE : PixelFormat.TRANSLUCENT;

            mRenderSurface = new SurfaceTextureHandler(surfaceTexture);

            surfaceTexture.setOnFrameAvailableListener(listener);

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.ICE_CREAM_SANDWICH_MR1) {
                if (mDesiredWidth > 0 && mDesiredHeight > 0) {
                    surfaceTexture.setDefaultBufferSize(mDesiredWidth, mDesiredHeight);
                }
            }
            Surface surface = new Surface(surfaceTexture);
            createSwapChain(surface);
            mRenderCallback.onResized(width,height);
        }
    }`

And here is my RenderCallback 

`        if(mSwapChain != null){
            mEngine.destroySwapChain(mSwapChain);
        }
        mSwapChain = mEngine.createSwapChain(surface,SwapChain.CONFIG_TRANSPARENT);`

Then I found the onFrameAvailable of the SurfaceTexture was called just once.That's Why?

And also I tried to save the frame texture into an image like this 
`    private void saveTextureToImage(int textureID, int width, int height, String fileName) {
        int[] frameBuffer = new int[1];

        GLES32.glGenFramebuffers(1, frameBuffer, 0);

        GLES32.glBindFramebuffer(GLES32.GL_FRAMEBUFFER, frameBuffer[0]);
        GLES32.glFramebufferTexture2D(GLES32.GL_FRAMEBUFFER, GLES32.GL_COLOR_ATTACHMENT0, GLES32.GL_TEXTURE_2D, textureID, 0);

        GLES32.glDrawBuffers(1, new int[]{GLES32.GL_COLOR_ATTACHMENT0}, 0);

        if (GLES32.glCheckFramebufferStatus(GLES32.GL_FRAMEBUFFER) != GLES32.GL_FRAMEBUFFER_COMPLETE) {
            Log.e(TAG, "framebuffer not complete");
            return;
        }

        ByteBuffer rgbaBuf = ByteBuffer.allocateDirect(width * height * 4);
        rgbaBuf.position(0);
        GLES32.glReadPixels(0, 0, width, height, GLES32.GL_RGBA, GLES32.GL_UNSIGNED_BYTE, rgbaBuf);

        saveRgb2Bitmap(rgbaBuf, fileName, width, height);

        GLES32.glDeleteFramebuffers(1, IntBuffer.wrap(frameBuffer));
        GLES32.glBindFramebuffer(GLES32.GL_FRAMEBUFFER, 0);
        GLES32.glGetError();
    }`

I got the glCheckFramebufferStatus - "framebuffer not complete".Does it any suggetion?

**Comment by romainguy** [comment]
> Then I found the onFrameAvailable of the SurfaceTexture was called just once.That's Why?

Do you properly call beginFrame/render/endFrame in a loop? Also note that the listener should be installed by the consumer, not the producer.

**Reply by RockVampire** [reply]
Of course , I use the 'Choreographer' to called beginFrame/render/endFrame in the do frame loop.And I think the listener is initialized by the consumer.

**Comment by RockVampire** [comment]
Is the call of Renderer->setDisplayInfo necessary? As you know ,I'm rendering on a Surfacetexture without a display

**Comment by duy-maimanh** [comment]
Hello @RockVampire .Have you solved this problem? I am having a similar problem. Can you share some suggestions?

**Reply by RockVampire** [reply]
Hello , I 'm sorry for seeing this so late. 
I don't know your conditions ,so I'm not sure your problem is in the same case with me .What I can do is just tell you what's the key problem in my project , and how I solve them.
The key problem at last for me is the `SurfaceTexture` must bind with external texture . Unfortunately , there's only one  external texture support in OpenGL ES , and it was occupied by the Android Camera Process. 
My purpose is to mix the texture from Camera and the texture from filament. So I have to use the SurfaceTexture with camera.
Finally , I solve the problem unbinding external texture by `glBindOESExternalTexture`  when I'm using it in filament .
Other things should be may attention to is if you are using another glContext to treat the frames from the Android Camera ,you should init filament with shared context in the glContext thread.
What else is just like @romainguy said , use the SurfaceTexture just as a Surface.
Hope these will help you ! And Thanks for @romainguy again!

**Reply by duy-maimanh** [reply]
@RockVampire Thanks for your answer. I have resolved my issue.

**Reply by dandingol03** [reply]
@RockVampire hi, Is there any possibility for me to review your code? I am running an ai engine to detect face landmarks and use filament to render 3d virtual asset based on face center transform matrix. These two engine run on opengles context, and i want filament can output render result to an opengles texture which can be processed with some other 2d effect. However i can't make it worked. 
Here is what i have tried:
- create an opengles texture with GL_TEXTURE_2D
- wrap the opengles texture with SurfaceTexture, and use it to create Surface
- use Surface to create SwapChain
- create filament with same opengl context, and use it to render stuff
- glbind this opengles texture, do some rendering

**Comment by romainguy** [comment]
> Unfortunately , there's only one external texture support in OpenGL ES , and it was occupied by the Android Camera Process.

That's incorrect, you can have multiple `SurfaceTexture` instances.
